module_analysis/1_calc_module_composition.py

```diff
@@ -0,0 +1,283 @@
+            membership_df[membership_df["module id"] == module_id].index
+        )
+        genes_outside_module = list(set(all_genes).difference(genes_in_module))
+
+        # Get core and accessory in module
+        core_genes_in_module = [
+            gene for gene in genes_in_module if gene in core_genes_list
+        ]
+        acc_genes_in_module = [
+            gene for gene in genes_in_module if gene in acc_genes_list
+        ]
+
+        # Get core and accessory genes outside of module
+        core_genes_outside_module = [
+            gene for gene in genes_outside_module if gene in core_genes_list
+        ]
+        acc_genes_outside_module = [
+            gene for gene in genes_outside_module if gene in acc_genes_list
+        ]
+
+        # Check
+        assert len(all_genes) == len(core_genes_in_module) + len(
+            acc_genes_in_module
+        ) + len(core_genes_outside_module) + len(acc_genes_outside_module)
+        assert len(genes_in_module) == len(core_genes_in_module) + len(
+            acc_genes_in_module
+        )
+
+        # Make contingency table
+        # -----|inside module|outside module
+        # core | # genes     | # genes
+        # acc  | # genes     | # genes
+        observed_contingency_table = np.array(
+            [
+                [len(core_genes_in_module), len(core_genes_outside_module)],
+                [len(acc_genes_in_module), len(acc_genes_outside_module)],
+            ]
+        )
+
+        # H0: The probability that the gene is core is the same
+        # whether or not you're in the module or outside
+        # H1: The probability that a gene is core is higher or lower inside the module
+        # than outside the module
+        # Should I do 2 one-sided tests, one for testing if
+        odds_ratio, pval = scipy.stats.fisher_exact(observed_contingency_table)
+
+        # odds_ratio, pval_greater = scipy.stats.fisher_exact(
+        #    observed_contingency_table, alternative="greater"
+        # )
+        # odds_ratio, pval_less = scipy.stats.fisher_exact(
+        #    observed_contingency_table, alternative="less"
+        # )
+
+        # Fill in df
+        out_df.loc[module_id, "num core"] = len(core_genes_in_module)
+        out_df.loc[module_id, "num acc"] = len(acc_genes_in_module)
+        out_df.loc[module_id, "odds ratio"] = odds_ratio
+        out_df.loc[module_id, "p-value"] = pval
+
+        # Do we want to include p-value criteria?
+        # Given the small sizes of the modules I think this is why
+        # most of the findings are not significant
+        # Use bonferrroni corrected p-value
+        if odds_ratio > 1 and pval < 0.05 / len(all_genes):
+            out_df.loc[module_id, "module label"] = "mostly core"
+        elif odds_ratio < 1 and pval < 0.05 / len(all_genes):
+            out_df.loc[module_id, "module label"] = "mostly accessory"
+        else:
+            out_df.loc[module_id, "module label"] = "mixed"
```

Comment:
You could do the following (not here but something to keep in mind):
```python

data_rows = []
data_rows.append({
"module_id":module_id,
"num core":len(core_genes_in_module),
"num acc":len(acc_genes_in_module),
"odds ratio":odds_ratio,
"p-value":pval,
"module_label": (
"mostly core" if odds_ratio > 1 and pval < 0.05 / len(all_genes) else 
"mostly accessory" if odds_ratio < 1 and pval < 0.05 / len(all_genes) else
"mixed")
})

return pd.DataFrame.from_records(data_rows)
```
This builds the dataframe implicitly with each row as a dictionary and saves lines of code.

Reply:
Another thing. It be helpful if you could include the total number of core and accessory genes outside your module. (aka fill out the full contingency table for each row.)

Reply:
Nice! I do remember seeing this in your code, i like it!